### ISSUE TITLE
Bug: ESLint API cannot find eslint-recommended.js

### ISSUE THREAD

**Issue by daidodo** [eslint/eslint#15575]
### Environment

Node version: v16.10.0
npm version: 7.24.0
Local ESLint version: 8.8.0
Global ESLint version: N/A
Operating System: macOS Big Sur

### What parser are you using?

Default (Espree)

### What did you do?

To reproduce the issue:
* `git clone https://github.com/daidodo/eslint-api-issue.git`
* `cd eslint-api-issue && npm i && npx webpack`
* `node dist/main.js src/index.ts`

### What did you expect to happen?

I expect the output is the ESLint resolved config for the given file.
If you run `npx tsc && node dist/main.js src/index.ts`, you can get:
```
ESLint version: 8.8.0
eslint object: ESLint {}
Resolving ESLint config:
{
  env: {},
  globals: {},
  noInlineConfig: undefined,
  parser: null,
  parserOptions: {},
  plugins: [],
  reportUnusedDisableDirectives: undefined,
  rules: {
    'no-unused-vars': [ 'warn', [Object] ],
    'constructor-super': [ 'error' ],
    'for-direction': [ 'error' ],
    'getter-return': [ 'error' ],
    'no-async-promise-executor': [ 'error' ],
    'no-case-declarations': [ 'error' ],
    'no-class-assign': [ 'error' ],
    'no-compare-neg-zero': [ 'error' ],
    'no-cond-assign': [ 'error' ],
    'no-const-assign': [ 'error' ],
    'no-constant-condition': [ 'error' ],
    'no-control-regex': [ 'error' ],
    'no-debugger': [ 'error' ],
    'no-delete-var': [ 'error' ],
    'no-dupe-args': [ 'error' ],
    'no-dupe-class-members': [ 'error' ],
    'no-dupe-else-if': [ 'error' ],
    'no-dupe-keys': [ 'error' ],
    'no-duplicate-case': [ 'error' ],
    'no-empty': [ 'error' ],
    'no-empty-character-class': [ 'error' ],
    'no-empty-pattern': [ 'error' ],
    'no-ex-assign': [ 'error' ],
    'no-extra-boolean-cast': [ 'error' ],
    'no-extra-semi': [ 'error' ],
    'no-fallthrough': [ 'error' ],
    'no-func-assign': [ 'error' ],
    'no-global-assign': [ 'error' ],
    'no-import-assign': [ 'error' ],
    'no-inner-declarations': [ 'error' ],
    'no-invalid-regexp': [ 'error' ],
    'no-irregular-whitespace': [ 'error' ],
    'no-loss-of-precision': [ 'error' ],
    'no-misleading-character-class': [ 'error' ],
    'no-mixed-spaces-and-tabs': [ 'error' ],
    'no-new-symbol': [ 'error' ],
    'no-nonoctal-decimal-escape': [ 'error' ],
    'no-obj-calls': [ 'error' ],
    'no-octal': [ 'error' ],
    'no-prototype-builtins': [ 'error' ],
    'no-redeclare': [ 'error' ],
    'no-regex-spaces': [ 'error' ],
    'no-self-assign': [ 'error' ],
    'no-setter-return': [ 'error' ],
    'no-shadow-restricted-names': [ 'error' ],
    'no-sparse-arrays': [ 'error' ],
    'no-this-before-super': [ 'error' ],
    'no-undef': [ 'error' ],
    'no-unexpected-multiline': [ 'error' ],
    'no-unreachable': [ 'error' ],
    'no-unsafe-finally': [ 'error' ],
    'no-unsafe-negation': [ 'error' ],
    'no-unsafe-optional-chaining': [ 'error' ],
    'no-unused-labels': [ 'error' ],
    'no-useless-backreference': [ 'error' ],
    'no-useless-catch': [ 'error' ],
    'no-useless-escape': [ 'error' ],
    'no-with': [ 'error' ],
    'require-yield': [ 'error' ],
    'use-isnan': [ 'error' ],
    'valid-typeof': [ 'error' ]
  },
  settings: {},
  ignorePatterns: [ '**/xxx.js' ]
}
```

### What actually happened?

If webpacked, the output would be:
```
ESLint version: 8.8.0
eslint object: ESLint {}

Error: Cannot read config file: /Users/xxx/Desktop/private/projects/conf/eslint-recommended.js
Error: Cannot find module '/Users/xxx/Desktop/private/projects/conf/eslint-recommended.js'
Require stack:
- /Users/xxx/Desktop/private/projects/eslint-api-issue/dist/noop.js
Referenced from: /Users/xxx/Desktop/private/projects/eslint-api-issue/package.json
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

I've submitted a similar issue (https://github.com/eslint/eslint/issues/13974) but got aged.
My guess is the `__dirname` in [lib/cli-engine/cli-engine.js](https://github.com/eslint/eslint/blame/896273f9d75c973ce2c7cc25580ae667a10ec6f9/lib/cli-engine/cli-engine.js#L580) causes the issue. But I don't have enough context to confirm or fix it.

Thanks!

**Comment by nzakas** [comment]
As we discussed in the original issue, ESLint isn’t designed to be bundled with Webpack, so this isn’t something the team can prioritize. If you want to experiment with fixing the problem and submit a pull request, we can review it.

**Comment by daidodo** [comment]
@nzakas, thanks for the response!

Could you share your thoughts about how it should be fixed?
I'm new to internal of ESLint so any guidance would help.

**Comment by daidodo** [comment]
I've dived a bit and found [this function](https://github.com/eslint/eslintrc/blob/f1179c587ae09fabb5c3402598363cfcec2494f7/lib/config-array-factory.js#L795) which loads config extends.
I think one solution is to cache the internal extend objects (`eslint:recommended` and `eslint:all`) instead of their file paths. Both ways the CLI (e.g. *eslint*) will get the same result.
What do you think? Thanks!

**Comment by nzakas** [comment]
The problem is that all config data is loaded via filepaths, so we can’t just swap in an object without changing a lot of the complex plumbing of the config cascade.

**Comment by daidodo** [comment]
Correct me if I'm wrong.

There are two parts of the changes:
1. Setting extend objects to the cache. That can be done by changing all the places that extend file paths are set. At a glance, those places are:
    - https://github.com/eslint/eslint/blob/5d60812d440762dff72420714273c714c4c5d074/lib/cli-engine/cli-engine.js#L619-L620
    - https://github.com/eslint/eslint/blob/9a4ae3b68a1afd9483d331997635727fb19a1a99/lib/cli-engine/file-enumerator.js#L218-L219
    - [eslintrc: flat-compat.js](https://github.com/eslint/eslintrc/blob/bdce01aad747393d13cd1ae17d8280e54c9c5191/lib/flat-compat.js#L228-L230)
    - [eslintrc: cascading-config-array-factory.js](https://github.com/eslint/eslintrc/blob/bdce01aad747393d13cd1ae17d8280e54c9c5191/tests/lib/cascading-config-array-factory.js#L41-L42)
    - [eslintrc: config-array-factory.js](https://github.com/eslint/eslintrc/blob/c5d49193a30b97e8c6bb2ad6306ee816afe8b3cc/tests/lib/config-array-factory.js#L37-L39)
2. Changing the way extend config is calculated. That's to update [_loadExtendedBuiltInConfig](https://github.com/eslint/eslintrc/blob/f1179c587ae09fabb5c3402598363cfcec2494f7/lib/config-array-factory.js#L795) as follow:
```js
    // ...
        if (extendName === "eslint:recommended") {
            return this._normalizeConfigData(this.eslintRecommendedConfig, {
                ...ctx,
                name: `${ctx.name} » ${extendName}`
            });
        }
    // ...
```

Plus there are changes to `class ConfigArrayFactory` constructor, of course.

Do you think if it would work? Thanks!

**Comment by nzakas** [comment]
I don’t know. Why don’t you try it and see what happens?

**Comment by mdjermanovic** [comment]
Reopening as this isn't fixed until https://github.com/eslint/eslint/pull/15616 gets merged.
